### PR TITLE
mail list route

### PR DESCRIPTION
**Pull request by MaiUbeid**
#27

### CODE REVIEW

File: server/app.js [hunk 1]
```diff
@@ -27,4 +28,9 @@ app.use((_req, res) => {
   res.sendFile(join(__dirname, '..', 'client', 'build', 'index.html'));
 });
 
+// eslint-disable-next-line no-unused-vars
+app.use((err, req, res, _next) => {
+  res.send({ statusCode: err.statusCode, error: errorHandle(err.statusCode) });
```
Comment (ahmedisam99) [inline server/app.js]: What if the error didn't have a `statusCode`? then the status code that'll be sent won't be `500`
I think it's better if you do the work for handling the error here, and no need for `errorHandle` middleware

File: server/controllers/mailList/index.js [hunk 2]
```diff
@@ -0,0 +1,30 @@
+const {
+  mailList: { addMail, getEmails },
+} = require('../../models/queries/');
+const { findEmail } = require('../../helpers/findEmail');
+const { validateEmail } = require('../../validation');
+
+exports.addEmail = async (req, res, next) => {
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]: if you are only exporting one thing then use `module.exports`

File: server/controllers/mailList/index.js [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+  const { email } = req.query;
+  try {
+    await validateEmail.validate({ email });
+    const allEmails = await getEmails();
+    const isExist = findEmail(allEmails, email);
+    if (isExist) {
+      res
+        .status(200)
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]: redundant
```suggestion
        
```

File: server/controllers/mailList/index.js [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        .send({ statusCdde: 200, message: 'Email is exist', data: email });
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]:
```suggestion
        .send({ statusCode: 200, message: 'Email already exists', data: { email } });
```

File: server/controllers/mailList/index.js [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    } else {
+      await addMail(email);
+      res.status(201).send({
+        statusCdde: 201,
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]:
```suggestion
        statusCode: 201,
```

File: server/controllers/mailList/index.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        message: 'Email is added successfully',
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]:
```suggestion
        message: 'Email was added successfully',
```

File: server/controllers/mailList/index.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+        data: email,
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]:
```suggestion
        data: { email },
```

File: server/controllers/middlewars/index.js [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+  let message = '';
+  switch (statusCode) {
+    case 413:
+      message = 'You exceeded the maximum allowed payload size.';
+      break;
+    default:
+      message = 'Internal Server Error';
+  }
+  return message;
+};
```
Comment (ahmedisam99) [inline server/controllers/middlewars/index.js]: redundant, just do it in the app (`app.js`)

File: server/helpers/findEmail.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+exports.findEmail = (allEmails, email) => {
```
Comment (ahmedisam99) [inline server/helpers/findEmail.js]: if only exporting one thing, use `module.exports`

File: server/helpers/findEmail.js [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+  let flag = false;
+  allEmails.forEach(element => {
```
Comment (ahmedisam99) [inline server/helpers/findEmail.js]: `Array.find` would be helpful here
Or you can simply use `Array.includes` directly in the controller w/o the need of this helper

File: server/models/queries/mailList.js [hunk 11]
```diff
@@ -0,0 +1,18 @@
+const base = require('../config');
+
+exports.addMail = email => {
+  return base('mailing_list').create([{ fields: { email } }]);
+};
+
+exports.getEmails = async () => {
+  let records = [];
+  const processPage = (partialRecords, fetchNextPage) => {
```
Comment (ahmedisam99) [inline server/models/queries/mailList.js]:
```suggestion
 
   const processPage = (partialRecords, fetchNextPage) => {
```

File: server/models/queries/mailList.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+    records = [...records, ...partialRecords.map(r => r.get('email'))];
+    fetchNextPage();
+  };
+  await base('mailing_list')
```
Comment (ahmedisam99) [inline server/models/queries/mailList.js]:
```suggestion
  
  await base('mailing_list')
```

File: server/validation/validateEmail.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+const yup = require('yup');
+
+exports.validateEmail = yup.object({
```
Comment (ahmedisam99) [inline server/validation/validateEmail.js]:
```suggestion
module.exports = yup.object({
```

File: test/mailList.test.js [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      } else {
+        t.equals(
+          res.text.includes('Invalid email address'),
+          true,
+          'Should have validation error',
+        );
+        t.end();
+      }
+    });
+});
```
Comment (ahmedisam99) [inline test/mailList.test.js]: how is this test passing?

File: server/controllers/mailList/index.js (hunk 2 above)
Comment (MaiUbeid) [reply server/controllers/mailList/index.js]: Ok

Comment (MaiUbeid) [reply server/controllers/mailList/index.js]: OK

File: server/controllers/mailList/index.js [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+      await addMail(email);
+      res.status(201).send({
+        statusCode: 201,
+        message: 'Email is added successfully',
+        data: email,
+      });
+    }
+  } catch (err) {
+    if (err.name === 'ValidationError')
+      res.status(400).send({ statusCdde: 400, error: 'Invalid email address' });
```
Comment (ahmedisam99) [inline server/controllers/mailList/index.js]:
```suggestion
      res.status(400).send({ statusCode: 400, error: 'Invalid email address' });
```

File: server/controllers/middleware/index.js [hunk 16]
```diff
@@ -0,0 +1,13 @@
+// eslint-disable-next-line no-unused-vars
+module.exports = (err, req, res, _next) => {
+  const { statusCode } = err;
```
Comment (ahmedisam99) [inline server/controllers/middleware/index.js]: add a `console.log` here for development

File: server/models/queries/mailList.js [hunk 17]
```diff
@@ -0,0 +1,20 @@
...
+exports.getEmails = async () => {
+  let records = [];
+
+  const processPage = (partialRecords, fetchNextPage) => {
+    records = [...records, ...partialRecords.map(r => r.get('email'))];
+    fetchNextPage();
+  };
+
+  await base('mailing_list')
+    .select({ view: 'Grid view', pageSize: 2 })
```
Comment (ahmedisam99) [inline server/models/queries/mailList.js]: why did you specify a page size of 2? 
```suggestion
    .select({ view: 'Grid view' })
```

File: server/validation/validateEmail.js [hunk 18]
```diff
@@ -0,0 +1,8 @@
+const yup = require('yup');
+
+module.exports = yup.object({
+  email: yup
+    .string()
+    .email()
+    .required(),
+});
```
Comment (ahmedisam99) [inline server/validation/validateEmail.js]: you could've simply used `yup.string.ema....` directly without using `yup.object`

**Pull request merged**